### DISCUSSION TITLE
How to turn off the screen on shutdown?

### DISCUSSION THREAD

**Discussion by crimsonnight** [mathoudebine/turing-smart-screen-python#123] [Q&A] [answered]
The screen is staying on after my PC has shut down (displaying the last frame). I've set it up to run at start-up via the task scheduler as recommended in the setup guide. Also, there is no tray icon as mentioned in the guide so cannot try exiting that way.

**Comment by vinzify** [comment]
On Win11, the tray icon appears as it should to me, but even when I close it, the screen remains on, last frame as for you.
Wish somone can help more

**Comment by mathoudebine** [comment]
For the screen to turn off when the program stops, you can do this manually in `main.py` just before the program stops: 
```python
display.lcd.ScreenOff()
```
Add the following code at the beginning of the `clean_stop()` function: https://github.com/mathoudebine/turing-smart-screen-python/blob/main/main.py#L36

You display should be turned off when the program exits cleanly: this is the case when stopped from the tray icon, but I don't know if this works when stopped on computer shutdown.
Please let me know.

If the screen does not turn on automatically when program restarts, you can add:
```python
self.lcd.ScreenOn()
```
here: https://github.com/mathoudebine/turing-smart-screen-python/blob/main/library/display.py#L68

**Reply by crimsonnight** [reply]
Thanks for answering, appreciate your time!

The weird thing is, I don't have a tray icon. Also, checking through the main.py you linked, it's pretty different to the one that's in the folder by default so I've updated my main.py with the contents of yours but still no tray icon. I've also added the line of code in as suggested, but when ending the task through the Task Scheduler the screen still stays on. Just to confirm, that section now looks like this:

`    def clean_stop(tray_icon=None):
        # Do not stop the program now in case data transmission was in progress
        # Instead, ask the scheduler to empty the action queue before stopping
        display.lcd.ScreenOff()
        scheduler.STOPPING = True`

**Reply by mathoudebine** [reply]
For the tray icon, you will need other modifications than just `main.py`. You can get the latest complete release (all the files) from the repository and apply your changes again if you want to have the tray icon.

I just tested the modification for the ScreenOff and it seems it only works when stopping from tray icon. When the program is stopped by the Task Scheduler, it is force-killed so I'm afraid there is no way to catch this event from the program to turn off the screen.

**Reply by Cyruz143** [reply]
Could do something to catch the shutdown event in Windows to kill it gracefully? Doesn't fix force kill but probably catches the most common use case. User shutsdown machine and doesn't manually close applications.

**Reply by mathoudebine** [reply]
Good idea indeed, maybe by using `win32api.setConsoleCtrlHandler()` to receive shutdown/logoff events from Windows then do a clean exit, it might be worth a try.
I also commited the modifications to turn off screen on a clean exit, they are already working when using tray icon / Ctrl+C

**Reply by crimsonnight** [reply]
> For the tray icon, you will need other modifications than just `main.py`. You can get the latest complete release (all the files) from the repository and apply your changes again if you want to have the tray icon.
> 
> I just tested the modification for the ScreenOff and it seems it only works when stopping from tray icon. When the program is stopped by the Task Scheduler, it is force-killed so I'm afraid there is no way to catch this event from the program to turn off the screen.

Hey :) I tried this but the program stopped working when I replaced my files with the updated ones (screen stays off when running the task). I tried rolling back the files one by one but I couldn't find what was causing the issue unfortunately. I've attached a zip of my working setup though - if you don't mind, would you mind taking a quick look and letting me know what might be causing the issue when trying to update? https://fromsmash.com/turningsetup

**Reply by mathoudebine** [reply]
It seems you started from the 2.2.0 release and you have not modified any Python code so the update should be quite easy:
- get the latest release (2.2.1)
- extract it in a specific folder
- edit the `config.yaml` to restore your configuration from your 2.2.0
- copy your new themes `1080papaya/`, `Landscape6GridCustomKami/` and `Landscape6GridCustomSango/` in `res/themes` from your 2.2.0  to the 2.2.1 folder
- don't forget your file `Landscape6GridCustom.psd` in `res/themes`

Everything should be working fine.
If you do not change Python code, it is always easier for an update to get the latest release, then edit `config.yaml` and copy your themes

**Reply by mathoudebine** [reply]
New feature request opened for this: #132

**Reply by crimsonnight** [reply]
Hm, I've done exactly that and it's no longer booting up - if I revert the scheduled task to use the previous directory it still works as expected. I could be doing something wrong, but I'm confident I've followed the outlined steps. I've uploaded the directory if you wanted to have a gander: https://fromsmash.com/turningsetup2
Definitely looking forward to the auto shut-down feature! Thanks :)

**Comment by mathoudebine** [comment, marked as answer]
Hello all, this feature has now been included in [Release 2.3.1](https://github.com/mathoudebine/turing-smart-screen-python/releases/tag/2.3.1)  
The screen will turn off automatically when program stops, or on computer logoff/shutdown

**Reply by crimsonnight** [reply]
Awesome, thank you!! I'm still only able to get 2.2.0 working for some reason though. I've triple-checked my configs are the same, tried with both the default theme and my custom theme, but I can't get the screen to turn on with anything other than 2.2.0. I've attached both directories, would you mind please letting me know what I'm doing wrong? https://fromsmash.com/turningss

**Reply by mathoudebine** [reply]
I downloaded your files and i can start the 2.3.2 successfully, so the issue may come from your config.yaml file. Can you open a terminal in the 2.3.2 folder, and run the main.py with python from this terminal? What is the output of the command?

**Reply by crimsonnight** [reply]
Thank you for checking it out - I've just upgraded my laptop so I've reinstalled Python, copied everything over and the new version is working fine - no idea what was happening previously, but thanks again for your assistance :)
